signature/src/signatures/sigstore.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::image;
+use anyhow::{anyhow, Result};
+use oci_distribution::Reference;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs;
+
+// FIXME: Now, for Demo, the dir which the sigstore config files locate
+// is hardcoded to "/etc/containers/registries.d".
+const REGISTRY_CONFIG_DIR: &str = "/etc/containers/registries.d";
+
+// Format the sigstore name:
+// `image-repository@digest-algo=digest-value`
+#[allow(unused_assignments)]
+pub fn format_sigstore_name(image_ref: &Reference, image_digest: image::digest::Digest) -> String {
+    let image_name = image_ref.repository().to_string();
+    format!(
+        "{}@{}={}",
+        image_name, image_digest.get_algorithm(), image_digest.get_value()
+    )
+}
+
+// Defines sigstore locations (sigstore base url) for a single namespace.
+#[derive(Serialize, Deserialize, Default, Debug, PartialEq, Clone)]
+struct SigstoreConfig {
+    #[serde(rename = "default-docker")]
+    default_docker: Option<RegistryNamespace>,
+    // The key is a namespace, using fully-expanded Docker reference format or parent namespaces.
+    docker: Option<HashMap<String, RegistryNamespace>>,
+}
+
+impl SigstoreConfig {
+    pub fn is_config_entry_exit(&self, ns: &str) -> bool {
+        if let Some(docker) = &self.docker {
+            return docker.get(ns).is_some();
+        } else {
+            false
+        }
+    }
+
+    pub fn insert_a_docker_config_entry(&mut self, ns_name: &str, ns_config: &RegistryNamespace) {
+        if let Some(docker) = &mut self.docker {
+            docker.insert(ns_name.to_string(), ns_config.clone());
+        } else {
+            let mut new_docker = HashMap::new();
+            new_docker.insert(ns_name.to_string(), ns_config.clone());
+            self.docker = Some(new_docker)
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+struct RegistryNamespace {
+    sigstore: String,
+}
+
+// loads sigstore configuration files(.yaml files) in specific dir.
+fn load_sigstore_config(dir: &str) -> Result<SigstoreConfig> {
+    let mut merged_config = SigstoreConfig::default();
+    let yaml_extension = OsStr::new("yaml");
+
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_dir() || path.extension() != Some(yaml_extension) {
+            continue;
+        }
+        let path_str = path
+            .to_str()
+            .ok_or(anyhow!("Unknown error: path parsed failed."))?;
+        let config_yaml_string = fs::read_to_string(path_str)?;
+        let config = serde_yaml::from_str::<SigstoreConfig>(&config_yaml_string)?;
```

Comment:
Just wanted to check if the intention is to really stop the loading of the sigstore files if one of them is a malformed YAML.

Reply:
What happens if there is a `sigstore-staging` entry like below? Does it ignore the entry or fail to parse?

```
              docker:
                  registry.example.com:
                      sigstore: https://registry-sigstore.example.com
                  registry.example.com/mydepartment:
                      sigstore: https://sigstore.mydepartment.example.com
                      sigstore-staging: file:///mnt/mydepartment/sigstore-staging
                  registry.example.com/mydepartment/myproject:mybranch:
                      sigstore: http://localhost:4242/sigstore
                      sigstore-staging: file:///home/useraccount/webroot/sigstore
```

Reply:
> Just wanted to check if the intention is to really stop the loading of the sigstore files if one of them is a malformed YAML.

This will only leads to the failure when the file content does not conform to the yaml format standard and the required fields are missing, which naturally leads to the failure of signature verification because signature cannot be obtained.


> What happens if there is a `sigstore-staging` entry like below? Does it ignore the entry or fail to parse?
> 
> ```
>               docker:
>                   registry.example.com:
>                       sigstore: https://registry-sigstore.example.com
>                   registry.example.com/mydepartment:
>                       sigstore: https://sigstore.mydepartment.example.com
>                       sigstore-staging: file:///mnt/mydepartment/sigstore-staging
>                   registry.example.com/mydepartment/myproject:mybranch:
>                       sigstore: http://localhost:4242/sigstore
>                       sigstore-staging: file:///home/useraccount/webroot/sigstore
> ```

`sigstore-staging` is the sigstore url only for writing. CC doesn't need to generete signatures so this filed is not needed, if there is a `sigstore-staging` entry, our code won't crash, but will ignore it.